README.md
# sclust
simple sentence clusterer

A tool to cluster lines of text in a streaming fashion.

E.g., if you have a file like this, with one sentence per line:

```
$ cat /tmp/foo
Hi there, how are you?
hi where how you are
i like to sing
I am going to sing
hi where how you are
hi there how...
do you sing???
```

You can pipe it to `sclust` to get cluster assignments:

```
$ cat /tmp/foo | sclust
0	Hi there, how are you?	-
0	hi where how you are	0.352689
1	i like to sing	-
1	I am going to sing	0.401597
0	hi where how you are	0.82069
0	hi there how...	0.45287
2	do you sing???	-
```
Here, the first column is the cluster assignment and the third column is the cosine similarity between that line and the cluster it is assigned to. The algorithm is online (as opposed to batch), so order matters. At each iteration, a document is assigned to its closest cluster, according to cosine similarity.

Other options:
```
$ sclust --help
A command-line tool to quickly cluster sentences.

usage:
    sclust [--help --threshold <T> --update-norms <N>]

Options
    -h, --help
    -t, --threshold <N>    Similarity threshold in [0,1]. Higher means sentences must be more similar to be merged. [default: .2]
    -u, --update-norms <N>    Update cluster norms every N documents. Larger values reduce run-time, but sacrifice accuracy. [default: 1]
```

There is also a tool `sclust-summarize` to view the output.

$ sclust-summarize --help
A command-line tool to print the top clusters output by sclust in a streaming fashion.
E.g., cat data.txt | sclust | sclust-summarize

```
usage:
    sclust-summarize [--help --frequency <F> --num-docs-to-print <N> --num-clusters-to-print <K>]

Options
    -h, --help
    -f, --frequency <F>               Print clusters every F lines [default: 1000]
    -n, --num-clusters-to-print <N>   Number of top clusters to print [default: 10]
    -k, --num-docs-to-print <K>       Number of documents per cluster to print [default: 1]
```

E.g.,

```
$ cat /tmp/foo | sclust | sclust-summarize  -k 3

---------7 documents, 3 clusters---------

4	0	hi where how you are	0.352689
 	 	hi where how you are	0.82069
 	 	hi there how...	0.45287
2	1	i like to sing	-
 	 	I am going to sing	0.401597
1	2	do you sing???	-
```
Here, the first column is cluster frequency, the second column is cluster id. You can optionally print the most recent `k` documents added to the cluster.

The nice thing is you can pipe a large file and see the clusters change as documents are read. E.g., 

```
$  cat /tmp/a_bunch_of_tweets  | sclust -t .2 -u 1000 | sclust-summarize -k 2 -n 5 

---------1000 documents, 715 clusters---------

15	68	lol	0.73664
 	 	i need matching jewelry me lj lol	0.215373
9	93	if you love watching huge zits getting popped then you re gonna love this it like therapeutic for me webaddress	0.273418
 	 	i love jesus exclamationpoint	0.314281
9	324	no one going to pay for content what they used to pay doesn t matter what it is	0.29804
 	 	what	0.646988
9	14	studing exclamationpoint studing exclamationpoint oh sh	0.23397
 	 	oh yeah this woman knows what she talkin about exclamationpoint	0.232043
7	160	rt	0.615143
 	 	rt plz rt fast waystation needs help evacuating animals exclamationpoint now exclamationpoint little tujunga canyon rt	0.269342

---------2000 documents, 1209 clusters---------

21	98	comendo happyemoticon	0.281718
 	 	we are better now and for good this time happyemoticon i love her happyemoticon	0.378463
19	68	lol who da hell u talkin bout	0.202604
 	 	see lol	0.448003
17	93	i love it here	0.512499
 	 	i love target and my lovelies	0.254959
16	14	oh yeah exclamationpoint lol	0.566714
 	 	uh oh	0.391978
15	135	lol i love hey arnold	0.642059
 	 	hey diggy	0.357601

---------3000 documents, 1621 clusters---------

29	93	i love you baby sister	0.349325
 	 	i love you diva o	0.366829
28	98	and im luvin it happyemoticon	0.263552
 	 	heyy everyone happyemoticon how are ya	0.225831
27	68	lol well i txt her just put jessica hahaha	0.207772
 	 	my bad i was falling myself lol	0.262167
25	135	and ahh exclamationpoint i love hey arnold exclamationpoint	0.599761
 	 	hey happyemoticon its exclamationpoint	0.53119
21	14	oh psych how you make my day	0.335252
 	 	oh my god oh my god oh my god	0.556064
```


HISTORY.rst
=======
History
=======

0.1.0 (2016-04-01)
------------------

* First release on PyPI.


tests/test_sclust.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_sclust
----------------------------------

Tests for `sclust` module.
"""

import unittest

from sclust import sclust


class TestSclust(unittest.TestCase):

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def test_000_something(self):
        pass


if __name__ == '__main__':
    import sys
    sys.exit(unittest.main())


sclust/sclust.py
# -*- coding: utf-8 -*-
"""A command-line tool to quickly cluster sentences.

usage:
    sclust [--help --threshold <T> --update-norms <N>]

Options
    -h, --help
    -t, --threshold <N>    Similarity threshold in [0,1]. Higher means sentences must be more similar to be merged. [default: .2]
    -u, --update-norms <N>    Update cluster norms every N documents. Larger values reduce run-time, but sacrifice accuracy. [default: 1]
"""
from collections import Counter, defaultdict
from docopt import docopt
from math import sqrt, log10
import numpy as np
import re
import sys


def norm(tokens, idfs):
    return sqrt(np.sum((count * idfs[token])**2
                       for token, count in tokens.items()))

def norm_dict(tokens, idfs):
    d = defaultdict(lambda: 0)
    denom = 0
    for token, count in tokens.items():
        v = count * idfs[token]
        d[token] = v
        denom += v*v
    denom = sqrt(denom)
    if denom == 0:
        denom = 1
    for token, val in d.items():
        d[token] = val / denom
    return d

def tokenize(line):
    return re.findall('\w+', line.lower())


def idf(token, doc_freqs, docnum):
    return log10((docnum + 1) / doc_freqs[token])

class Cluster:
    def __init__(self, token_counts):
        self.token_counts = Counter(token_counts)
        self.cluster_norm = None

    def add(self, tokens):
        self.token_counts.update(tokens)

    def score(self, normed_doc, idfs, do_update_norm):
        if do_update_norm or not self.cluster_norm:
            self.normed_cluster = norm_dict(self.token_counts, idfs)
        return sum(value * self.normed_cluster[token] for token, value in normed_doc.items())

def update_index(index, clusterid, tokens):
    for t in tokens:
        index[t].add(clusterid)

def search_index(index, top_words):
    clusters = set()
    for w in top_words:
        clusters |= index[w]
    return clusters

def run(threshold, norm_update):
    doc_freqs = Counter()
    clusters = []
    docnum = 0
    index = defaultdict(set)
    for line in sys.stdin:
        docnum += 1
        do_update_norm = True if docnum % norm_update == 0 else False
        line = line.strip()
        tokens = Counter(tokenize(line))
        if len(tokens) == 0:
            continue
        doc_freqs.update(tokens)
        idfs = {token: idf(token, doc_freqs, docnum) for token, value in doc_freqs.items()}
        normed_doc = norm_dict(tokens, idfs)
        # What are the five words with highest tfidf weight? Use to filter comparisons.
        top_words = sorted(tokens, key=lambda x: -normed_doc[x])[:5]  # (doc_freqs[token], token) for token in tokens)
        best_cluster = -1
        best_score = -1
        for ci in search_index(index, top_words):
            cluster = clusters[ci]
            score = cluster.score(normed_doc, idfs, do_update_norm)
            if score > best_score and score > threshold:
                best_cluster = ci
                best_score = score
        if best_cluster == -1:
            clusters.append(Cluster(tokens))
            update_index(index, len(clusters)-1, tokens)
            print('%d\t%s\t-' % (len(clusters)-1, line))
        else:
            clusters[best_cluster].add(tokens)
            update_index(index, best_cluster, tokens)
            print('%d\t%s\t%g' % (best_cluster, line, best_score))
        sys.stdout.flush()


# Weirdness when piping to unix tools. See http://stackoverflow.com/a/26736013/1756896
def _void_f(*args,**kwargs):
    pass

def main():
    args = docopt(__doc__)
    try:
        run(float(args['--threshold']),
            int(args['--update-norms']))
    except (BrokenPipeError, IOError):
        sys.stdout.write = _void_f
        sys.stdout.flush = _void_f
        sys.exit()



if __name__ == '__main__':
    main()



travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file
"""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except:
    from urllib.request import urlopen


GITHUB_REPO = 'aronwc/sclust'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key, with work-around for keys using
    incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning.
    """
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Update the deploy section of the .travis.yml file
    to use the given encrypted password.
    """
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


sclust/sclust_summarize.py
# -*- coding: utf-8 -*-
"""A command-line tool to print the top clusters output by sclust in a streaming fashion.
E.g., cat data.txt | sclust | sclust-summarize

usage:
    sclust-summarize [--help --frequency <F> --num-docs-to-print <N> --num-clusters-to-print <K>]

Options
    -h, --help
    -f, --frequency <F>               Print clusters every F lines [default: 1000]
    -n, --num-clusters-to-print <N>   Number of top clusters to print [default: 10]
    -k, --num-docs-to-print <K>       Number of documents per cluster to print [default: 1]
"""
from collections import Counter, defaultdict, deque
from docopt import docopt
from math import sqrt, log10
import numpy as np
import re
import sys

def print_summary(cluster_counts, clusterid2lines, num_clusters, lineno):
    print('\n---------%d documents, %d clusters---------\n' % (lineno, len(cluster_counts)))
    for clusterid, count in cluster_counts.most_common(num_clusters):
        lines = clusterid2lines[clusterid]
        print('%d\t%s' % (count, lines[0]))
        for line in list(lines)[1:]:
            print(' \t \t%s' % '\t'.join(line.split('\t')[1:]))
    sys.stdout.flush()

def run(frequency, num_clusters, num_docs):
    cluster_counts = Counter()
    clusterid2lines = defaultdict(lambda: deque(maxlen=num_docs))
    lineno = 0
    for line in sys.stdin:
        line = line.strip()
        lineno += 1
        fields = line.split('\t')
        cluster_counts[fields[0]] += 1
        clusterid2lines[fields[0]].append(line)
        if lineno % frequency == 0:
            print_summary(cluster_counts, clusterid2lines, num_clusters, lineno)
    print_summary(cluster_counts, clusterid2lines, num_clusters, lineno)


# Weirdness when piping to unix tools. See http://stackoverflow.com/a/26736013/1756896
def _void_f(*args,**kwargs):
    pass

def main():
    args = docopt(__doc__)
    try:
        run(int(args['--frequency']), int(args['--num-clusters-to-print']),
            int(args['--num-docs-to-print']))
    except (BrokenPipeError, IOError):
        sys.stdout.write = _void_f
        sys.stdout.flush = _void_f
        sys.exit()



if __name__ == '__main__':
    main()

